### PR TITLE
Rework bitenum impl & move tests

### PR DESCRIPTION
**Pull request by nicopap**
Rewrite the bitenum module to use more rusty patterns.

1. Use the native syn parsing stuff to parse the attributes arguments
2. Remove all instances of panic! with syn::Result. This results in much higher quality error messages, highlighting precisely the bits of code that cause the error.
4. Allow arbitrary order in specification of the uN and 'exhaustive' arguments
5. Move logic in different functions/methods. This reduces the complexity of the `bitenum` function, also reduces the indentation level.
6. Add a regression test for `exhaustive` elision in `bitenum` macro call

<details><summary>Examples of actionable error messages</summary>

Note: the macro always suggest as discriminant value the largest encountered discriminant + 1.

```
error: The enum has 3 variants, it would need 4 variants to be exhaustive. Either add variants, use a smaller storage type, or mark this enum as 'exhaustive = false'.
  --> bitbybit-tests/src/bitenum_tests.rs:26:32
   |
26 |     #[bitenum(u2, exhaustive = true)]
   |                                ^^^^
```
```
error: The enum has 4 variants, it is exhaustive for 2 bits. Either remove variants, use a larger storage type, or mark this enum as 'exhaustive = true'.
  --> bitbybit-tests/src/bitenum_tests.rs:59:32
   |
59 |     #[bitenum(u2, exhaustive = false)]
   |                                ^^^^^
```

```
error: Discriminants must be literal integers. Eg: 'Two = 2' '0x2' '0b10' '0o2'.
  --> bitbybit-tests/src/bitenum_tests.rs:63:15
   |
63 |         Two = 1 << 1,
   |               ^^^^^^
```

```
error: The largest discriminant value is larger than can be stored in the provided storage type. Max discriminant value is 127.
  --> bitbybit-tests/src/bitenum_tests.rs:80:15
   |
80 |         One = 255,
   |               ^^^
```

```
error: All variants of a #[bitenum] must have an explicit literal discriminant. Eg: 'Three = 3'.
  --> bitbybit-tests/src/bitenum_tests.rs:82:9
   |
82 |         Three,
   |         ^^^^^
```

</details>

I'm still a bit dissatisfied with the way enum discriminants are parsed into integers. I'm not exactly sure it's necessary…

I've been looking at the bitfield module, and I'm a bit terrified, I'm opening this PR before starting work on the other one. This way, I avoid doing work that you would refuse to merge outright.

Why move the tests inside the bitbybit-tests package? Because it enables using the -Zunpretty rustc flag to inspect the expanded macros, I used the following commands:

```
RUSTC_BOOTSTRAP=1 cargo rustc --tests -p bitbybit-tests -- --cfg=tests -Zunpretty=expanded
```
I also added `resolver = 2` to the workspace `Cargo.toml`, `cargo` kept nagging me about it. Also set `publish = false` in `bitbybit-tests` so that it isn't published by accident.

### CODE REVIEW

**Review by danlehmann — COMMENTED** [review]
Wow, thank you so much for the cleanup! I'll make sure to review this in depth shortly

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:10]: Size can be up to u64

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:18]:
```suggestion
    "Discriminants must be literal integers, either binary, hexadecimal, octal or decimal";
```

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:41]: remove the "use" here? Can just fully qualify in the line below

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:122]: please use /// for doc comments

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:213]: The name 'constructor' is a bit misleading. Without context I'd expect a constructor for Self here (pretty much what new_with_raw_value() is).

Maybe call it raw_value_constructor?

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:189]: fyi, I had an outstanding patch to fix the Err-Type here: https://github.com/danlehmann/bitfield/pull/47

But as that's a breaking change, I held off from doing this one until a major version bump

**Review by danlehmann — COMMENTED** [review]
Ok, debugged through it all. Excellent refactor, thanks a lot!

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:70]: A if/else does really the same thing, but I prefer `match ()` in this particular case. I find it more readable.

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:41]: It's only so that rustfmt doesn't expand the following if/else into several lines. I'll swap back to use the qualified path

**Comment by nicopap** [comment]
Thanks for the review, I will put this PR in draft mode until I address all the issues (and do a bit of additional testing)

**Comment by danlehmann** [comment]
With regards to opening up the syntax to allowing arbitrary ordering, it would be good to add a test that uses the opposite order to make sure a future patch doesn't break it (something like `#[bitenum(exhaustive = false, u64)`).

Don't worry about adding more tests to the _legacy test suite - restricting it to the new one is fine.

**Comment by monkeydom** [comment]
Just as a philosophical point in terms of API: Do we really want the order to be arbitrary? From an API standpoint I kind of like the fact that the first argument needs to be the type and explicit. To my taste I think we want the first argument to have to be a type and to be present.

**Review by monkeydom — CHANGES_REQUESTED** [review]
Overall fantastic work, thanks!

**Comment by monkeydom** [comment]
I do think we should also add tests for the default values for exhaustive if we want to keep the existing behavior.

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:70]: This is purely subjective. I'm happy to defer to anyone's preference. I'll swap to if/else

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:126]: I'm not exactly sure we can parse it the same way. <https://docs.rs/syn/latest/syn/struct.LitInt.html> Seems to not support non-base 10 literals. Am I reading this right?

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:93]: I've no preference in any particular direction. Just beware that ensuring an ordering requires an additional check and error message. It's basically adding an `if self.explicit_bits.is_none() return Err…` in the "exhaustive" branch.

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:178]: I see. But I think we need to adjust the error message.

**Comment by nicopap** [comment]
I put back the default fallback to `exhaustive = false`. This would have been a breaking change otherwise.

I'm for removing the default in a future potentially breaking release though.

File: bitbybit/src/bit_size.rs
Comment (nicopap) [reply bitbybit/src/bit_size.rs:34]: The way it should be done is:

1. Have separates crate for the proc macro and the `bitbybit` crate.
2. re-export both the macros and `arbitrary_int` in `bitbybit`
3. Use `::bitbybit::arbitrary_int::…` instead of `arbitrary_int` in the macro generation code.
4. Ideally, add an additional attribute argument to `bitenum` that let end-users customize the `::bitbybit`. `serde` and `enumset` are crates that allow this.
5. Finally, fully qualify the std lib types used in the macro (so it would be `::core::result::Result` instead of `Result`)

File: bitbybit/src/bitenum.rs
Comment (nicopap) [inline bitbybit/src/bitenum.rs:40]: This could be replaced by a `thiserror::Error` derive macro. But it does add a fairly slow-to-compile dependency, which isn't ideal when used in a proc macro.

File: bitbybit/src/bitenum.rs
Comment (monkeydom) [inline bitbybit/src/bitenum.rs:134]: This needs to return an error on size > 64, otherwise we just panic for u128 and bigger before returning an error due to the shift left later.

**Review by monkeydom — COMMENTED** [review]
Great turnaround time, good adjustments. Thanks!

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:171]: Actually not! We can test this in the playground. I'll swap to use then. Nice! <https://play.rust-lang.org/?gist=aa881a12e57778322e018a296fdfe946>

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:48]: I'm 100% for this. Also we could drop having to explicitly declare `conditional`. I think we should write up all the API evolutions we listed in this PR discussions in an issue to keep track of them.

File: bitbybit/src/bitenum.rs
Comment (nicopap) [reply bitbybit/src/bitenum.rs:134]: It's already the case. I'm getting the following error:

```
error: The specified storage size is invalid, valid is in range 1..=64
  --> bitbybit-tests/src/bitenum_tests.rs:26:15
   |
26 |     #[bitenum(u67)]
   |               ^^^
```

File: bitbybit/Cargo.toml
Comment (danlehmann) [inline bitbybit/Cargo.toml:21]: why this change?

the current version of proc-macro2 for example is "1.0.69", so this change would move us back in time

File: bitbybit/src/bit_size.rs
Comment (danlehmann) [reply bitbybit/src/bit_size.rs:34]: Why change the current system? Dependency too loosely specified?

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [reply bitbybit/src/bitenum.rs]: Either way is fine with me. I like having the data type as a first argument, but if someone really wants to swap them I think that shouldn't be a problem

File: bitbybit/src/bitenum.rs
Comment (danlehmann) [inline bitbybit/src/bitenum.rs:232]: that's a great idea, never occurred to me!

File: bitbybit/src/bit_size.rs
Comment (nicopap) [reply bitbybit/src/bit_size.rs:34]: The current system makes the macro "unhygienic". As in: it pulls identifiers from some surrounding scope, which is pretty poor.

Specifically for `Result`, consider someone importing a custom `Result` type like `use syn::Result`. It would result in a compilation error, since now the macro would refer to `syn::Result` when using `Result`.

For `arbitrary_int`, the issue is that we have to add it as an explicit dependency. Which is pretty poor user experience. There is also problems about keeping the two versions in sync. If we re-imported it ourselves, it would shift the responsibility to us, not to our many users.

File: bitbybit/Cargo.toml
Comment (nicopap) [reply bitbybit/Cargo.toml:21]: That's an error on my part.

Context: The third party crates.io registry called `lib.rs` has a "lint" page on dependencies (see yours here <https://lib.rs/~danlehmann/dash>). I found on my own lint page that it is better to specify the whole version string, due to potential violation of semantic versioning (there are at least two well known cases in the rust ecosystem).

But, at the time, I was specifying just the major version number, not even the minor one.

I think you are right, specifying the patch number is superfluous.

File: bitbybit/src/bitenum.rs
Comment (monkeydom) [reply bitbybit/src/bitenum.rs]: Okay, then lets have it as is and be positional independent but required.

File: bitbybit/src/bit_size.rs
Comment (danlehmann) [reply bitbybit/src/bit_size.rs:34]: Agree about Result! Fully qualifying it would certainly fix it.

For arbitrary-int, I'm less sure. You certainly have a point, but I think it would also be annoying for users to deal with having two namespaces of arbitrary-int, which likely are mostly the same.

**Review by danlehmann — COMMENTED** [review]
This looks good to me now. I'll have to do one more big pass (and test it with two larger projects I am working on), but I think everything has been addressed. I'll merge once I'm done with that.

Thanks again for your contribution!

**Review by danlehmann — APPROVED** [review]

**Comment by danlehmann** [comment]
Thanks again for your contribution!

**Pull request merged**
